Glow pro 2024 with serial data

Rongrrz opened the pull request:
Major revamp to GlowPro2024 project
- Deleted a lot of unnecessary stuff
- Added lots of comments and documentation
- Updated pattern functions to display well and properly
- Renamed "Light" class to "LightStrip" (for readability)

Rongrrz committed:
Big revamp to the system, added serial data retrieval

Rongrrz committed:
Deleted .vscode\settings.json

Rongrrz committed:
Reset tolerance count if data

Rongrrz committed:
Properly instantiate lights

Rongrrz committed:
Slight changes

Rongrrz committed:
Type hinting changes (Won't affect deploy)

aschokking commented on .vscode/settings.json:
⭐ If you want this file to stay gone, add it to the `.gitignore` file too

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
-        light.primary = Colors["BLUE"]
-        light.pattern = patterns.static
+[33] Disabled with auto
+RoboRIO is disabled, abd dose have autonomous program set
```
⭐ nit: typo in comment abd dose

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
+            tolerance_count = 0
+            return data.strip()
+    
+    # Check tolerance and return previous, if tolerance count reached, "31"
```
⭐ could you elaborate on this comment about what it means for tolerance to be reached? eg, that the robot has failed to send data for a while? how long in s?

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
+            tolerance_count = 0
+            return data.strip()
+    
+    # Check tolerance and return previous, if tolerance count reached, "31"
```
Also were you able to test this? eg stop the robot program and see it kick in?

Rongrrz replied:
We tested this using another pull request under GlowPro2024 (do not merge) and it was able to receive serial data.

aschokking replied:
I mean the opposite, that if the serial data stops coming that it reverts to no code

Rongrrz replied:
I don't think we have tested that before 😅 💦

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
+
+
+# Assign colors and patterns to light strips
+def assign_mode(light: LightStrip):
```
⭐ low priority for later, but I'm curious if this shouldn't just be a method on the LightStrip class where you pass in the modes

Rongrrz replied:
Didn't think of that 😅

Rongrrz replied:
Tried to implement it but then saw that it'd take a lot more effort than expectedd. assign_mode right now uses current_mode inside the main_loop.py file, which also imports lightstrip, then if lightstrip file wants to use current_mode... well, wouldn't that make it error out? (Because you need to import main_loop)

aschokking replied:
Right you'd pass in the current_mode and prev_mode as arguements into the method

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
+    light.pattern_starting_loop = patterns.loopcount
+
+    # Update color and pattern on light to display
+    if current_mode == "1":
```
⭐ ⭐ because it can be so easy to forget what mode is what, instead of using strings create some named const variables a the top of the file like:
```
AMP_SIGNAL = "1"
```

And then this code can become:
```
    if current_mode == AMP_SIGNAL :
```

Which I think is much easier to read

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
         pass
+    else:
+        # If serial data received does not match any of our mode
+        # Or is 31 (no code)
```
These are all TODOs to put some kind of pattern in here right?

Rongrrz replied:
I'll add a comment for these :> (And in fact they are to-dos)

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
-    elif light.state == 3:
-        light.primary = Colors["GRAY"]
-        light.pattern = patterns.static
+--< Modes for when RoboRIO is enabled >--
```
⭐ ⭐ ⭐ are you missing a crucial state for robot is enabled but none of these 4 states are active?

Rongrrz replied:
Never thought about that before... I'll add that

aschokking commented on main.py:
```diff
@@ -1,93 +1,120 @@
-        main_loop(lights)
-
+        main_loop()
+        time.sleep(patterns.frames_per_second / 1)
```
⭐ ⭐ ⭐ this doesn't make a lot of sense to me, frames per second is going to be some big number like 20 fps right? This would sleep for 20s in that case?

aschokking replied:
Also I think it would make more sense for this to live inside main_loop() either way. In Python its usually better to have as little code inside your top level `if __name__ == "__main__":` block as possible, ideally just calling your main function

Rongrrz replied:
I made a mistake with the time.sleep, it was meant to be 1/fps

I also moved the if __name__ == "__main__" to another file :)

aschokking reviewed (approved):
The codes on a great trajectory, let's try and implement at least some placeholder state for for all the remaining states so we can start testing things end to end (even if we decide to change the particular color/pattern later)

j0ndough commented on main.py:
```diff
@@ -1,93 +1,120 @@
+# Retrieves and returns serial data sent by Java side
+def get_serial_data() -> str:
+    # Return serial data if any
+    if supervisor.serial_bytes_available > 0:
```
nit: in Python, if a value is nonzero and you evaluate it in an `if` statement, it will evaluate to true, so the `> 0` part should be unnecessary.

Rongrrz committed:
Resolved comments

Rongrrz committed:
Organized modes into a dictionary

Rongrrz committed:
Minor organization adjustment

Rongrrz committed:
Fixed time.sleep()

Rongrrz committed:
Once again, reorganized main

I don't want to do this ever again.

Rongrrz committed:
Filled in remaining patterns

Also we probably need to talk with drive team about patterns :).

aschokking commented on main_loop.py:
```diff
@@ -0,0 +1,150 @@
+serial = sys.stdin
+tolerance = 2
+tolerance_count = 0
+current_mode = "31"
```
⭐ ⭐ instead of hardcoding these to strings, if you created named variables for these it would be so much nicer to read this as:
```python
current_mode = NO_CODE
```

aschokking replied:
or seeing the code you have below it could be
```python
current_mode = modes["NO_CODE"]

aschokking commented on main_loop.py:
```diff
@@ -0,0 +1,150 @@
+
+
+# MODES
+modes = {
```
⭐ ⭐ Missing no code?

aschokking commented on lightstrip.py:
```diff
@@ -0,0 +1,21 @@
+        self.secondary: tuple[int, int, int] = (0, 0, 0)
+
+        # Pattern function to call on every update
+        self.pattern = static
```
⭐ because it can be confusing using functions as variables, suggest adding a unit to this one so its `pattern_func`

Rongrrz committed:
Resolved PR comments

Moved modes and assign_mode method to lightstrip.py alongside other minor changes for clarity.
